src/coprocessor/dag/dag.rs

```diff
@@ -63,52 +67,98 @@ impl DAGContext {
             exec: dag_executor.exec,
             output_offsets: req.take_output_offsets(),
             batch_row_limit: batch_row_limit,
+            chunks_per_stream: chunks_per_stream,
+            chunks: Vec::new(),
         })
     }
 
-    pub fn handle_request(&mut self) -> Result<Response> {
+    pub fn handle_request(&mut self, streaming: bool) -> Result<(Response, bool)> {
         let mut record_cnt = 0;
-        let mut chunks = Vec::new();
+        let (mut first_row, mut start_key) = (true, None);
         loop {
             match self.exec.next() {
                 Ok(Some(row)) => {
                     self.req_ctx.check_if_outdated()?;
-                    if chunks.is_empty() || record_cnt >= self.batch_row_limit {
-                        let chunk = Chunk::new();
-                        chunks.push(chunk);
+
+                    if first_row {
+                        first_row = false;
+                        start_key = self.exec.take_last_key();
+                    }
+
+                    let mut stream_result = None;
+                    if self.chunks.is_empty() || record_cnt >= self.batch_row_limit {
+                        if streaming && self.chunks.len() >= self.chunks_per_stream {
+                            let start_key = start_key.take();
+                            let end_key = self.exec.take_last_key();
+                            stream_result = Some(self.make_response(true, start_key, end_key));
```

Comment:
I prefer to return `stream_result` here since it won't change once generated.

Reply:
There are some other works to do before return.